src/Mvc/Mvc.Core/src/ApplicationModels/AuthorizationApplicationModelProvider.cs

```diff
@@ -74,16 +74,25 @@ public void OnProvidersExecuting(ApplicationModelProviderContext context)
 
     public static AuthorizeFilter GetFilter(IAuthorizationPolicyProvider policyProvider, IEnumerable<IAuthorizeData> authData)
     {
-        // The default policy provider will make the same policy for given input, so make it only once.
-        // This will always execute synchronously.
-        if (policyProvider.GetType() == typeof(DefaultAuthorizationPolicyProvider))
+        if (GetPolicyIfDefaultProvider(policyProvider, authData) is AuthorizationPolicy policy)
         {
-            var policy = AuthorizationPolicy.CombineAsync(policyProvider, authData).GetAwaiter().GetResult()!;
             return new AuthorizeFilter(policy);
         }
         else
         {
             return new AuthorizeFilter(policyProvider, authData);
         }
     }
+
+    // The default policy provider will make the same policy for given input, so make it only once.
+    // This will always execute synchronously.
+    public static AuthorizationPolicy? GetPolicyIfDefaultProvider(IAuthorizationPolicyProvider policyProvider, IEnumerable<IAuthorizeData> authData)
+    {
+        if (policyProvider.GetType() != typeof(DefaultAuthorizationPolicyProvider))
+        {
+            return null;
+        }
+
+        return AuthorizationPolicy.CombineAsync(policyProvider, authData).GetAwaiter().GetResult();
```

Comment:
Is there another place to do this that is async?

Reply:
Yes. We're forced to redo this logic in every call into `AuthorizeFilter.OnAuthorizationAsync(context)` for the non-`DefaultAuthorizationPolicyProvider` case since non-default providers might really be async. This was existing behavior.

This PR doesn't change when async operations occur, but it does mutate the `AuthorizeFilter` so we can continue using the same instance to keep the overridden behavior. This mutation is merely a caching optimization if the same service provider is always passed to `IFilterFactory.CreateInstance(IServiceProvider)` but could lead to some weird bugs otherwise.

Reply:
You know how I feel about sync over async here... If there's someplace we can do it that's async, then we should.